### DISCUSSION TITLE
uiMidiKeyboard always on top?

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#609] [Q&A] [answered]
I would like to put label components on top (above) of the uiMidiKeyboard, but it looks like they always stay behind the keyboard. Is this true and if so, is it possible to add code to change this?

**Comment by dnaldoog** [comment]
Hi @bijlevel,

You should be able to achieve that using layers. Put the uiMidiComponent in a layer below the labels you want to superimpose over it. 

[Here](https://ctrlr.org/forums/topic/using-layers-tutorial-1-0/) is an old tutorial I wrote about it on the old website.

[LAYERS tut.01.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/14273937/LAYERS.tut.01.bpanelz.zip)

**Comment by unityconstruct** [comment]
@bijlevel : 
- I just started a panel teardown for XStation that (i'm my limited experience) seems to make effective use of layers for simplicity (ie: less hassle). 
- Main thing to note is level0 is UNDER level5... which might be super obvious, until you have to account for the layerGUI presenting 5 below 0. 
- Not a huge deal, once you know about it, but might having one chasing ghosts.

https://github.com/unityconstruct/CtrlrDocs/blob/master/ui/xstation/ui-design/xstation-ui-design-guide.md

- other item of importance: when to show/hide.. might be arbitrary that layers[4+5} are shown by default, then hidden on afterPanelLoaded()....
- things _might_ not load if it layer isn't visible, but haven't seen anything suggesting it other than XStation implementation
```lua
function afterPanelLoaded()
	panelFullyLoaded = 0
	showGreeting()
	showHideExtraControls(false)
end
```

@dnaldoog ... stealing your notes at that link... 


@ both: Sharing is caring: https://github.com/unityconstruct/CtrlrPanels

**Reply by bijlevel** [reply]
Thanks for the input. I know how to work with layers, but it is AFAIK not the solution. I have attached a screenshot of what I want to achieve. The various colours indicate the range of assigned samples (in this case) to the keyboard, but I also want the user to be able to click on the keyboard the hear the sample sound. But because the keyboard is in a layer behind the coloured labels, it is not active, so does not react to mouse clicks, and as soon as I make the keyboard layer active, the colours disappear, of course...

<img width="1222" alt="Screenshot 2024-02-14 at 11 39 36" src="https://github.com/RomanKubiak/ctrlr/assets/11859363/f084543e-1a93-41ad-81c8-1f8fdca73bc0">

**Reply by dnaldoog** [reply]
In that case you could create a lot of fillRect objects in a uiComponent and a onMouseUp routine perhaps 🤔

**Reply by bijlevel** [reply]
You mean imitating the keyboard with fillRect objects? BTW the range does change when a user assigns different ranges to the samples. I think it is all doable, but seems an awful lot of (coding) work! I'm aiming at the most simple solution 😉!

**Reply by bijlevel** [reply]
Maybe this is easier, just keep the colours away from the keyboard:

<img width="1217" alt="Screenshot 2024-02-14 at 12 42 36" src="https://github.com/RomanKubiak/ctrlr/assets/11859363/65602bae-5466-4f44-82e0-c3d90b558357">

**Reply by unityconstruct** [reply]
@bijlevel  Village Idiot here again....🤤
1. a workaround might be to set the height of the `keymap regions` to 1/2 that of the keyboard, or above it
    - providing visual/tactile access to both simultaneously.
    - On the panel I'm looking at currently, enabling a `higher` layer to show some extra knob-age doesn't interfere with controls still visible.
2. ~~a more complicated implementation might be to use a `mouseDown() event` & `CtrlrPanelCanvas:getX()` with a table look up to determine notes..~~ but... well,  meh....
3. last idea is to set transparency on the midi keyboard keys & have keymapping shapes on a lower layer with higher saturation so they don't get washed out.
    - CtrlrMidiKeyboard::CtrlrMidiKeyboard

```lua
CtrlrMidiKeyboard::CtrlrMidiKeyboard (CtrlrModulator &owner)
    : CtrlrComponent(owner),
      midiKeyboard (0)
{
    addAndMakeVisible (midiKeyboard = new MidiKeyboardComponent (keyboardState, MidiKeyboardComponent::horizontalKeyboard ));
    midiKeyboard->setName ("midiKeyboard");
	keyboardState.addListener (this);
	setProperty (Ids::uiMidiKeyboardOrientation, "horizontalKeyboard");
	setProperty (Ids::uiMidiKeyboardWhiteButtonColour, "0xffffffff");
	setProperty (Ids::uiMidiKeyboardBlackButtonColour, "0xff000000");
	setProperty (Ids::uiMidiKeyboardSeparatorLineColour, "0xff000000");
	setProperty (Ids::uiMidiKeyboardMouseOverColour, "0xffff0000");
	setProperty (Ids::uiMidiKeyboardMouseDownColour, "0xff0000ff");
	setProperty (Ids::uiMidiKeyboardTextLabelColour, "0xff000000");
	setProperty (Ids::uiMidiKeyboardButtonBackgroundColour, "0xff0f0f0f");
	setProperty (Ids::uiMidiKeyboardButtonArrowColour, "0xff000000");
	setProperty (Ids::uiMidiKeyboardLowestVisibleKey, 48);
	setProperty (Ids::uiMidiKeyboardBaseOctaveKeyPress, 0);
	setProperty (Ids::uiMidiKeyboardOctaveFroMiddleC, 3);
	setProperty (Ids::uiMidiKeyboardMapToNoteNumber, false);
...
```

![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/c12fc210-eccd-405a-b4ce-3d08f678ab87)

_...but don't let me mislead you that I have any idea what i'm talking about... as I assure you that I don't_

**Reply by bijlevel** [reply, marked as answer]
@unityconstruct idea 3 works perfect! (Most of the time the best ideas come from the insane and the idiots😋😋! )And thanks to the colourpicker of Damien I can easily play with the amount of transparancy! Still a lot of coding to do though, because the ranges and the number of keymaps can vary.

<img width="1211" alt="Screenshot 2024-02-14 at 14 05 07" src="https://github.com/RomanKubiak/ctrlr/assets/11859363/dba623ba-1a19-45f8-80ff-102defdda23f">

**Reply by unityconstruct** [reply]
haha.. even stopped clock is right 2x daily.. Glad that 10days of bleeding-code-hunting-eyes weren't for naught.. 

LOOKS SEXY!

The `.29` color picker is bomb.. + over UI 'tightness'.. Damian should bump the version  to 9.1 out of pride

**Comment by bijlevel** [comment]
> @ both: Sharing is caring: https://github.com/unityconstruct/CtrlrPanels

@unityconstruct I had a look at the overview of the various panels on the old forum via your github. I see you've added our TX7 panel, but not the DX7 version. Could you please add that one or tell me how I can add it myself?

**Reply by unityconstruct** [reply]
too funny , my `ghetto bash wget for loop` picked up some things the `httrack crawler` missed... I want to injure the spammers on Ctrlr.org though.... sifted through like 40K folders to parse out most the nonsense.. TONS of good info on that site in any case

this the one or is there another?.. about to add it
![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/1d1896e6-a90d-4411-a002-293245e0d28f)

**Reply by bijlevel** [reply]
Yep, that's the one. Thanks!
